### PR TITLE
fix(hosting): hosting main services refactored

### PR DESCRIPTION
**Pull request by rfougier**

### CODE REVIEW

File: client/app/hosting/ftp/hosting-ftp.controller.js [hunk 1]
```diff
@@ -1,187 +1,191 @@
...
+            this.loadTab(count, offset, true);
+        }
 
-    $scope.$on(Hosting.events.tasksChanged, () => {
-        $scope.disableRestoreFtp = true;
-    });
+        loadTab (count, offset, needUsers) {
+            this.Hosting.getTabFTP(this.$stateParams.productId, count, offset, needUsers, this.search.value)
+                .then((ftpInformations) => {
+                    if (!_.isEmpty(ftpInformations.list.results)) {
```
Comment (Apollidore) [inline client/app/hosting/ftp/hosting-ftp.controller.js]: I would have said`if (ftpInformations  != null)` then we know that it has `list.results`

File: client/app/hosting/ftp/hosting-ftp.controller.js [hunk 2]
```diff
@@ -1,187 +1,191 @@
...
+            delete user.isPrimaryAccount;
+
+            this.HostingUser.updateUser(this.$stateParams.productId, {
+                login: user.login,
+                data: _.omit(user, "ssh", "ftp", "ftpUrl", "sshUrl")
+            }).then(() => {
+                this.Alerter.success(this.$scope.tr("hosting_tab_FTP_configuration_user_modify_success"), this.$scope.alerts.dashboard);
+                this.startPolling();
+            }).catch((err) => {
+                const idx = _.indexOf(this.ftpInformations.list.result, element);
```
Comment (Apollidore) [inline client/app/hosting/ftp/hosting-ftp.controller.js]: `_.find` ?

File: client/app/hosting/ftp/hosting-ftp.service.js [hunk 3]
```diff
@@ -1,125 +1,142 @@
...
+    class HostingUser {
+        /**
+         * Constructor
+         * @param Hosting
+         * @param OvhHttp
+         * @param Poller
+         * @param $rootScope
+         */
+        constructor (Hosting, OvhHttp, Poller, $rootScope) {
+            this.Hosting = Hosting;
```
Comment (Apollidore) [inline client/app/hosting/ftp/hosting-ftp.service.js]: `Object.assign(this, Hosting, OvhHttp, Poller, $rootScope)`

File: client/app/hosting/ftp/hosting-ftp.service.js [hunk 4]
```diff
@@ -1,125 +1,142 @@
...
+                interval: 5000,
+                successRule: {
+                    state (task) {
+                        return opts.successSates.indexOf(task.state) !== -1;
+                    }
+                },
+                namespace: opts.namespace
+            }).then((pollObject, task) => {
+                this.$rootScope.$broadcast(`${opts.namespace}.done`, pollObject, task);
+            }, (err) => {
```
Comment (Apollidore) [inline client/app/hosting/ftp/hosting-ftp.service.js]: `catch`

File: client/app/hosting/hosting-tabs.controller.js [hunk 5]
```diff
@@ -1,101 +1,121 @@
...
+            this.$scope = $scope;
+            this.$q = $q;
+            this.$location = $location;
+            this.$stateParams = $stateParams;
+            this.HostingFreedom = HostingFreedom;
+            this.HostingIndy = HostingIndy;
 
-    const defaultTab = "GENERAL_INFORMATIONS";
-    $scope.toKebabCase = _.kebabCase;
+            $scope.displayTabs = { cron: true, databases: true, modules: true };
```
Comment (Apollidore) [inline client/app/hosting/hosting-tabs.controller.js]: `$onInit`

File: client/app/hosting/hosting.service.js [hunk 6]
```diff
@@ -1,575 +1,646 @@
...
+             */
+            static isPerfOffer (offer) {
+                return /^perf.+$/.test(offer);
+            }
 
-    this.isPerfOffer = (offer) => /^perf.+$/.test(offer);
+            /**
+             * Is password valid
+             * @param {string} password
+             * @param {{min: *, max: *}|null} customConditions
```
Comment (Apollidore) [inline client/app/hosting/hosting.service.js]: `*` instead of `number`

File: client/app/hosting/hosting.service.js [hunk 7]
```diff
@@ -1,575 +1,646 @@
...
+             * @returns {boolean}
+             */
+            static isPasswordValid (password, customConditions = undefined) {
+                const min = _.get(customConditions, "min", passwordConditions.min);
+                const max = _.get(customConditions, "max", passwordConditions.max);
+                return !!(password && password.length >= min && password.length <= max && password.match(/.*[0-9].*/) && password.match(/.*[a-z].*/) && password.match(/.*[A-Z].*/) && password.match(/^[a-zA-Z0-9]+$/));
```
Comment (Apollidore) [inline client/app/hosting/hosting.service.js]: Could be simplified

File: client/app/hosting/hosting.service.js [hunk 8]
```diff
@@ -1,575 +1,646 @@
...
+            }
 
-    this.cloudWebUnlimitedQuantity = 100000;
+            /**
+             * Is path valid
+             * @param {string} path
+             * @returns {boolean}
+             */
+            static isPathValid (path) {
+                return /^[a-zA-Z0-9-._\/]*$/.test(path) && !/\.\./.test(path);
```
Comment (Apollidore) [inline client/app/hosting/hosting.service.js]: `a-zA-Z0-9_` => `\d`

File: client/app/hosting/hosting.service.js [hunk 9]
```diff
@@ -1,575 +1,646 @@
...
+            /**
+             * Get specific hosting service
+             * @param {string} serviceName
+             * @param {array|null} catchOpt
+             */
+            getHosting (serviceName, catchOpt) {
+                return this.OvhHttp.get(`/hosting/web/${serviceName}`, {
+                    rootPath: "apiv6"
+                }).then((data) => data).catch((http) => {
+                    if (catchOpt && _.isArray(catchOpt) && _.indexOf(catchOpt, http.status) !== -1) {
```
Comment (Apollidore) [inline client/app/hosting/hosting.service.js]: `!_.isEmpty(catchOpt)`

File: client/app/hosting/hosting.service.js [hunk 10]
```diff
@@ -1,575 +1,646 @@
...
-                        }
-                    })
-                    .then((response) => {
-                        if (response.data && response.data.length) {
-                            tasks = _.union(tasks, response.data);
-                        }
-                    })
-            );
-        });
+                }).then((resp) => resp, (err) => err)));
```
Comment (Apollidore) [inline client/app/hosting/hosting.service.js]: `catch`

File: client/app/hosting/hosting.service.js [hunk 11]
```diff
@@ -1,575 +1,646 @@
...
+            getAttachDomainSslLinked (serviceName) {
+                const defered = this.$q.defer();
+
+                this.OvhHttp.get(`/hosting/web/${serviceName}/ssl/domains`, {
+                    rootPath: "apiv6",
+                    returnSuccessKey: ""
+                }).then((resp) => {
+                    defered.resolve(resp.data);
+                }).catch((err) => {
+                    if (err.status === 404) {
```
Comment (Apollidore) [inline client/app/hosting/hosting.service.js]: OvhHttp doesn't give status anymore

File: client/app/hosting/hosting.service.js (hunk 8 above)
Comment (rfougier) [reply client/app/hosting/hosting.service.js]: @FredericEspiau a-zA-Z0-9_ => \w

Comment (FredericEspiau) [reply client/app/hosting/hosting.service.js]: @rfougier indeed :)

File: client/app/hosting/ftp/hosting-ftp.controller.js (hunk 2 above)
Comment (rfougier) [reply client/app/hosting/ftp/hosting-ftp.controller.js]: Really ?

File: client/app/hosting/hosting.service.js (hunk 7 above)
Comment (rfougier) [reply client/app/hosting/hosting.service.js]: We'll do that in future refactoring task

**Pull request merged**